File: 1169-invalid-transactions/1169-invalid-transactions.java
class Solution {
    public List<String> invalidTransactions(String[] transactions) {
        List<String> invalid = new ArrayList<>();
        HashMap<String, List<Transaction>> map = new HashMap<>();
        
        for( String transaction : transactions){
            Transaction t = new Transaction(transaction);
            map.putIfAbsent(t.name, new ArrayList<>());
            map.get(t.name).add(t);
        }
        
        for(String transaction : transactions){
            Transaction t = new Transaction(transaction);
            if(!isValid(map.get(t.name), t)){
                invalid.add(transaction);
            }
        }
        return invalid;
    }
    
    public boolean isValid(List<Transaction> transactions, Transaction transaction){
        if(transactions.size() <= 1 && transaction.amount < 1000)
            return true;
        
        for(Transaction t : transactions){
            if(transaction.inValidTransaction(t.city, t.time)){
                return false;
            }
        }
        
        return true;
    }
}

class Transaction{
    String name;
    int id;
    int time;
    int amount;
    String city;
    
    
    public Transaction(String t){
        String[] info = t.split(",");
        this.name = info[0];
        this.time = Integer.parseInt(info[1]);
        this.amount = Integer.parseInt(info[2]);
        this.city = info[3]; 
    }
    
    public boolean inValidTransaction(String city, int time){
        return invalidAmount() || differentCity(city, time);
    }
    
    public boolean invalidAmount(){
        return this.amount > 1000;
    }
    
    public boolean differentCity(String city, int time){
        return !this.city.equals(city) && Math.abs(this.time - time) <= 60;
    }
}

File: 1169-invalid-transactions/1169-invalid-transactions.py
from collections import defaultdict
class Solution:
    def invalidTransactions(self, transactions: List[str]) -> List[str]:
        t_map = defaultdict(list)
        invalid = []
        for t in transactions:
            txn = self.Transaction(t)
            t_map[txn.name].append(txn)
            
        for t in transactions:
            txn = self.Transaction(t)
            for t2 in t_map[txn.name]:
                if self.isInvalid(txn, t2):
                    invalid.append(t)
                    break
        
        return invalid
    def isInvalid(self, t1, t2):
        return t1.amt > 1000 or (t1.city != t2.city and abs(t1.time - t2.time) <= 60)
    
    class Transaction:
        def __init__(self, txn):
            self.txn = txn.split(",")
            self.time = int(self.txn[1])
            self.city = self.txn[3]
            self.amt = int(self.txn[2])
            self.name = self.txn[0]